apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/PODPipeline.ts

```diff
@@ -0,0 +1,439 @@
+        if (
+          atom.entries[atom.matchTo.entry].value.toString() ===
+          semaphoreId.toString()
+        ) {
+          matchingAtoms.push(atom);
+        }
+      }
+    }
+
+    return matchingAtoms;
+  }
+
+  /**
+   * Get a cached PCD for the given ID. IDs are derived from the POD
+   * entries and pipeline ID. We also want to ensure that the POD was signed
+   * with the current EdDSA private key, so we include that in the cache key.
+   *
+   * @param id The id of the PCD to get
+   * @returns The cached PCD, or undefined if it is not cached
+   */
+  private async getCachedPCD(id: string): Promise<SerializedPCD | undefined> {
+    const key = await getHash(`${id}${this.eddsaPrivateKey}`);
+    const cacheEntry = await this.cacheService.getValue(key);
+    if (cacheEntry) {
+      return JSON.parse(cacheEntry.cache_value);
+    }
+    return undefined;
+  }
+
+  /**
+   * Store a serialized PCD in the cache.
+   *
+   * @param id The id of the PCD to store
+   * @param serializedPCD The PCD to store
+   */
+  private async storeCachedPCD(
+    id: string,
+    serializedPCD: SerializedPCD
+  ): Promise<void> {
+    const key = await getHash(`${id}${this.eddsaPrivateKey}`);
+    await this.cacheService.setValue(key, JSON.stringify(serializedPCD));
+  }
+
+  /**
+   * Return a response to the feed issuance request. This takes the form of
+   * an array of PCD actions that will be applied to the user's PCD collection.
+   *
+   * @param req The request for feed issuance
+   * @returns The response to the feed issuance request
+   */
+  private async feedIssue(
+    req: PollFeedRequest
+  ): Promise<PollFeedResponseValue> {
+    return traced(LOG_NAME, "feedIssue", async (span) => {
+      if (!req.pcd) {
+        throw new Error("missing credential pcd");
+      }
+
+      if (this.definition.options.paused) {
+        return { actions: [] };
+      }
+
+      const credential =
+        await this.credentialSubservice.verifyAndExpectZupassEmail(req.pcd);
+      const { email, semaphoreId } = credential;
+
+      span?.setAttribute("email", email);
+      span?.setAttribute("semaphore_id", semaphoreId);
+
+      // Consumer is validated, so save them in the consumer list
+      const didUpdate = await this.consumerDB.save(
+        this.id,
+        email,
+        semaphoreId,
+        new Date()
+      );
+      if (didUpdate) {
+        this.consumerDB.save(this.id, email, semaphoreId, new Date());
+      }
+
+      const atomsToIssue = await this.loadMatchingAtoms(credential);
+
+      const serializedPCDs = await Promise.all(
+        atomsToIssue.map(async (atom) => {
+          // Get the final POD entries from the atom, output configuration,
+          // and credential
+          const entries = finalizeAtom(
+            atom,
+            this.definition.options.outputs[atom.outputId],
+            credential
+          );
+
+          const id = uuidv5(serializePODEntries(entries), this.id);
```

Comment:
See my discussion of stable IDs in atomize.ts, as it applies here.
The use of caching suggests you do want a stable content-based ID (hash or POD contentID).  It use as PCD ID returned to Zupass might suggest otherwise, if you care about doing replacements without deleting first.

Reply:
I'm now using `PODContent.fromEntries()` in order to take advantage of the key-sorting, but not using the `contentID` (for reasons discussed elsewhere).